keps/sig-storage/2268-non-graceful-shutdown/README.md

```diff
@@ -0,0 +1,466 @@
+}
+```
+
+* The Graceful Node Shutdown [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2000-graceful-node-shutdown) introduced a way to detect a node shutdown. When the NodeShutdownManager detects a node is shutting down, it sets the `Reason` field of a `PodStatus` to a constant named `nodeShutdownReason`. Currently `nodeShutdownReason` is in the `nodeshutdown` package. We need to export this constant by renaming it to `NodeShutdownReason` and move it to a package that can be imported by the Pod GC Controller.
+
+Existing logic:
+1. When a node is partitioned from the control plane, the health check in Node lifecycle controller, part of kube-controller-manager, sets Node v1.NodeReady Condition to False or Unknown (unreachable) if lease is not renewed for a specific grace period. Node Status becomes NotReady.
+
+1. After 300 seconds (default), the Taint Manager tries to delete Pods on the Node after detecting that the Node is NotReady. The Pods will be stuck in terminating status.
+
+Proposed logic change:
+1. [Proposed change] In the Pod GC Controller, part of the kube-controller-manager, add a new function called `gcTerminating`. This function would need to go through all the Pods in terminating state, verify that the Node is NotReady, check if the `Reason` in the `PodStatus` is set to the `NodeShutDownReason` constant, and then check if the `SafeDetach` flag in CSIDriver is set to true. If so, do the following:
+  1. [Proposed change] The Pod GC Controller would also apply a `quarantine` taint on the nodes whose pods are forcefully deleted due to the shutdown. This should happen before the pods are being evicted. The `quarantine` taint indicates that new pods should not be scheduled on the node.
+  1. [Proposed change] After applying taint on the shutdown node, forcefully delete pods that are stuck.
+
+1. [Proposed change] Once pods are selected and forcefully deleted, the attachdetach reconciler should check if the `SafeDetach` flag in CSIDriver is set to true. If so, it should skip the `attachedVolume.MountedByNode` check for the volumes backing these pods and allow `volumeAttachement` to be deleted.
+
+1. This would trigger the deletion of the `volumeAttachement` objects. This would allow `ControllerUnpublishVolume` to happen before `NodeUnpublishVolume` and/or `NodeUnstageVolume` are called.
+
+1. When the `external-attacher` detects the `volumeAttachement` object is being deleted, it calls CSI driver's `ControllerUnpublishVolume`.
+
+1. [Proposed change] The CSI driver would need to ensure that no volume is being used through a mount point. If unable to determine the usage, the CSI driver can return an error without detaching the volume. The `external-attacher` would then retry with backoffs.
+
+### Handle the Return of Shutdown Node
+
+If the nodes that were previously shutdown with pods forcefully deleted come back, we need to make sure they are cleaned up before pods are scheduled on them again.
```

Comment:
the scope of the KEP now is changed. It requires user to make sure node is down already which means no pods should run at this point.

Reply:
So if we can't CONFIRM it is powered off, we still let it hang, right?

Reply:
If the quarantine taint was not added to the shutdown node, the pods will be stuck in terminating status forever if the shutdown node remains down. If the shutdown node comes up, however, the original volumeAttachment will be deleted and the pods will be moved to a new node successfully. Added this to the KEP to clarify.

Reply:
I have tried if we use following configs in statefulset, the pods would be in stuck in terminating status for ~7min, what's the benefit if user already sets  following configs? Manually add a taint to a node in problem is still not an ideal way IMO:
```yaml
    spec:
      tolerations:
      - key: "node.kubernetes.io/not-ready"
        operator: "Exists"
        effect: "NoExecute"
        tolerationSeconds: 10
      - key: "node.kubernetes.io/unreachable"
        operator: "Exists"
        effect: "NoExecute"
        tolerationSeconds: 10
      terminationGracePeriodSeconds: 0
```
check details here: 
https://github.com/kubernetes-sigs/azuredisk-csi-driver/tree/master/docs/known-issues/node-shutdown-recovery

Reply:
@andyzhangx What K8s version have you tested this with?  Also which command did you use to shutdown the node?  Just want to check if this goes thru the Graceful Node Shutdown code path.

Reply:
@xing-yang tested with ~1.19, and I powered off the node directly.

Reply:
@andyzhangx I checked your link. In your workaround, why those tolerations are needed? When node is shutdown, pod is stuck in terminating and that's why disk will not be detached from the node. 
When setting terminationGracePeriodSeconds to 0,  will it make a difference so that pod will be  force deleted from API server even if the node is down (kubelet will not response)?

Reply:
@jingxu97 I tried again with `k apply -f https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/master/docs/known-issues/node-shutdown-recovery/statefulset-azuredisk.yaml` on k8s v1.20.9 cluster, after I power off node `aks-agentpool-35915375-vmss000000`, you could see `statefulset-azuredisk-0` is stuck in `Terminating` forever, while `statefulset-autorecover-0` would be terminated and scheduled to node `aks-agentpool-35915375-vmss000001` after a few minutes. I think `terminationGracePeriodSeconds: 0` makes the difference, with that setting, pod would be terminated immediately in pod eviction when node `aks-agentpool-35915375-vmss000000` becomes `NotReady`, I think you could get same test result.

So here the point is `terminationGracePeriodSeconds: 0` already have same effect, do we really need this KEP since this KEP requires adding a taint manually on the node in problem.

```console
# k get po -o wide
NAME                                    READY   STATUS              RESTARTS   AGE     IP             NODE                                NOMINATED NODE   READINESS GATES
statefulset-autorecover-0               1/1     Running             0          10m     10.240.0.130   aks-agentpool-35915375-vmss000001   <none>           <none>
statefulset-azuredisk-0                 1/1     Terminating         0          28m     10.240.0.35    aks-agentpool-35915375-vmss000000   <none>           <none>
```

Reply:
@andyzhang To set terminationGracePeriodSeconds, I think user need to add this to every pod spec. For system admin, they will not be able to do that.

Also in normal cases, we might not recommend to set terminationGracePeriodSeconds to 0 since this will not allow pod to gracefully terminate even on healthy node which might cause undesired behavior. 

The goal of this KEP is to recover form node shutdown automatically. As the first phase, we still require minimum human intervention, in this case, to add a taint to the node. In the future, this could be done by controller automatically.